pallene/assignment_conversion.lua

```diff
@@ -0,0 +1,349 @@
+        if stat.else_ then
+            self:visit_stat(stat.else_)
+        end
+    
+    elseif tag == "ast.Stat.ForNum" then
+        stat.start = self:convert_exp(stat.start)
+        stat.limit = self:convert_exp(stat.limit)
+        stat.step  = self:convert_exp(stat.step)
+
+        self:register_decl(stat.decl)
+        self:visit_stats(stat.block.stats)
+
+    elseif tag == "ast.Stat.ForIn" then
+        for _, decl in ipairs(stat.decls) do
+            self:register_decl(decl)
+        end
+
+        self:convert_exps_of_stat(stat)
+        self:visit_stats(stat.block.stats)
+
+    elseif tag == "ast.Stat.Assign" then
+        for i, var in ipairs(stat.vars) do
+            stat.vars[i] = self:convert_var(var)
+            
+            if var._tag == "ast.Var.Name" and not var._exported_as then
+                if var._def._tag == "checker.Def.Variable" then
+                    local decl = assert(var._def.decl)
+                    local depth = self.func_depth_of_decl[decl]
+                    local func_info = self.func_stack[depth]
+                    table.insert(func_info.mutated_decls, decl) 
+                end
+            end
+        end
+
+        self:convert_exps_of_stat(stat)
+
+    elseif tag == "ast.Stat.Decl" then
+        for _, decl in ipairs(stat.decls) do
+            self:register_decl(decl)
+        end
+
+        self:convert_exps_of_stat(stat)
+
+    elseif tag == "ast.Stat.Call" then
+        stat.call_exp = self:convert_exp(stat.call_exp)
+
+    elseif  tag == "ast.Stat.Break" then
+        -- empty
+    else
+        typedecl.tag_error(tag)
+    end
+end
+
+
+function Converter:convert_var(var)
+    local vtag = var._tag
+
+    if vtag == "ast.Var.Name" and not var._exported_as then
+        if var._def._tag == "checker.Def.Variable" then
+            local decl = assert(var._def.decl)
+            assert(self.ref_to_decl[decl], decl.name)
+            local depth = self.func_depth_of_decl[decl]
+            -- depth == 1 when the decl is that of a global
+            if depth < #self.func_stack and depth > 1 then
+                local func_info = self.func_stack[depth]
+                func_info.captured_decls[decl] = true
+            end
+            table.insert(self.ref_to_decl[decl], var)
+        end
+
+    elseif vtag == "ast.Var.Dot" then
+        var.exp = self:convert_exp(var.exp)
+    
+    elseif vtag == "ast.Var.Bracket" then
+        var.t = self:convert_exp(var.t)
+        var.k = self:convert_exp(var.k)
+    
+    end
+
+    return var
+end
+
+-- If necessary, transforms `exp` or one of it's subexpression nodes in case they 
+-- reference a mutable upvalue.
+-- Recursively visits all sub-expressions and applies an `ast.Var.Name => ast.Var.Dot`
+-- transformation wherever necessary, returning the new transformed node.
+-- Note that for the transformed node to be updated at the call site, this method must
+-- be called like so:
+-- `exp = self:convert_exp(exp)`
```

Comment:
Do we ever return an `exp` that is not the same that we received as an argument? Maybe I'm missing it but I can't see where that happens. If this really does not happen, then we could return nothing instead.

I also ask the same question about the `convert_var`.

Reply:
You're right. This was a misjudgement on my end. I changed that and the names to`visit_exp` and `visit_var`